readme

ePages shell, Beta 2

Simple command-line tool to work with the epages objects.

This experimental tool aims to include in one shell all the operations you need to work with 
epages at low level: manage objects (create, read, update, delete) and more... 

epages already includes a lot of scripts to do almost everything, but it would be nice to 
have those utilities in one tool. 

INSTALL

    Simply set the execution rights to the launcher script
    
        # cd epages-shell
        # chmod 755 she    

USAGE

    she [ flags ]  [ options ] [ StoreName [ ObjectPath ] ]

    Flags (optional):
        -help           Shows this help
        -noheader       Don't print the starting help header
        -nopager        Disable pager
        -prompt         Enable print prompt in Batch mode 
        
    Options:
        -batch          Batch mode: execute specified text file (optional)

    Arguments:
        Storename       Connect to the StoreName (optional)
        ObjectPath      And select the object by path (optional)

    Examples:
        she 
        she Store /Shops/DemoShop
        she -nopager -noheader Site
        she -prompt -batch script.txt

RUN

    Try 
    
        # she --help
        
    for command-line options  
    
    Simple execution
    
        # she

    Or you can start she with arguments to connect a store:
        
        # she Store
    
        # she Store /Shops/DemoShop

BATCH MODE

  You can run she with a batch file, no user interaction !
  
  Try the sample batch file:
  
    # she sample.batch
    
FIRSTS STEPS

    0. List of available commands and help for a specific one:
        
        [she] help
        [she] ? set
        
    1. Connect to a Store or BusinessUnit (database):
    
        [she] use Store
        
    2. Browse the objects tree
    
        [she] cd /Shops/DemoShop
        [she] cd Users
        
    3. Information about the current object
    
        [she] status
        
    4. List the childs of the current object
     
        [she] ls
        
    5. List all the attributes of the current object
    
        [she] get
        
    6. Change the Birth date of a customer
     
        [she] cd /Shops/DemoShop/Customers/1000/BillingAddress
        [she] set Birthday='21/08/1989 09:30:00'

TO DO

  Uf !!! Lots of things ...



history.txt

HISTORY : ePages shell

Release a.b

    Feature 1
    
    Fixed 1
    
    Fixed 2
    
Release a.b

    Feature 1
    
    Fixed 1
    
    Fixed 2


lib/ePages/Shell.pm
#======================================================================================================================
# Shell
#======================================================================================================================
package ePages::Shell ;
use base Shell::Shell ;

use strict ;

use DE_EPAGES::Database::API::Connection qw ( 
    RunOnStore 
) ;
use DE_EPAGES::Core::API::Error qw (
        ExistsError
        GetError
    ) ;
use ePages::ePages ;

#======================================================================================================================
# §function     new
# §state        public
#----------------------------------------------------------------------------------------------------------------------
# §syntax       Shell->new() ;
#----------------------------------------------------------------------------------------------------------------------
# §description  Constructor
#----------------------------------------------------------------------------------------------------------------------
# §return       $Object | The new object instance | Object
#======================================================================================================================
sub new {
    my $class = shift;

    my $hOptions = $_[0] // {} ;

    my $hAttributes  = {
        'Name'              => 'she',
        'Title'             => 'SHE: Shell ePages (Beta 2)',
        'Prompt'            => '[she] ',
        'ResetStore'        => 0,
        'ePages'            => ePages::ePages->new(), 
        'Commands'          => [
            'ePages/Command/Store',
            'ePages/Command/Path',
            'ePages/Command/Status',
            'ePages/Command/Childs',
            'ePages/Command/Get',
            'ePages/Command/Set',
            'ePages/Command/Delete',
            'Shell/Command/Test',
        ],
    } ;


    return $class->SUPER::new( { %$hAttributes, %$hOptions } );
}

#======================================================================================================================
# §function     getHeaderText
# §state        public
#----------------------------------------------------------------------------------------------------------------------
# §syntax       pending
#----------------------------------------------------------------------------------------------------------------------
# §description  pending
#======================================================================================================================
sub getHeaderText {
    my $self = shift;

    return <<HELP_TEXT

Wellcome to the ePages shell, a simple command-line tool to ease the interaction with the epages objects.

This is an experimental tool (Alpha version) which intends to include in one shell all the operations you need
to work with epages at low level: manage objects (create, read, update, delete) and more... Yes, epages includes
a lot os scripts to do almost everything, but it would be nice to have those utilities in one tool, isn't ? 

First steps:
    0. List of available commands and help for a specific one:
        >> help
        >> ? set
    1. Connect to a Store or BusinessUnit (database):
        >> use Store
    2. Browse the objects tree
        >> cd /Shops/DemoShop
        >> cd Users
    3. Information about the current object
        >> status
    4. List the childs of the current object 
        >> ls
    5. List all the attributes of the current object
        >> get
    6. Change the Birth date of a customer 
        >> cd /Shops/DemoShop/Customers/1000/BillingAddress
        >> set Birthday='21/08/1989 09:30:00'

Command-line arguments: she [ StoreName ] [ ObjectPath ]
    
    You can start she with arguments to connect a store:
        # she Store
    and select an object:
        # she Store /Shops/DemoShop
        
HELP_TEXT
}

#======================================================================================================================
# §function     _run
# §state        public
#----------------------------------------------------------------------------------------------------------------------
# §syntax       $Shell->_run()
#----------------------------------------------------------------------------------------------------------------------
# §description  TODO
#======================================================================================================================
sub _run {
    my $self = shift ;

    my $Arguments = $self->getArguments() ;
    my $StoreName = $Arguments->{'@'}->[0] ; 
    if ( defined $StoreName ) {
        $self->executeCommand( 'use', $StoreName ) ;
    }
        
    do {
        $self->{'ResetStore'} = 0 ;
        if ( defined $self->{'ePages'}->getStore() ) {
            $self->runOnStore() ;
        } else {
            $self->SUPER::_run() ;
        }
    } while ( $self->{'ResetStore'} ) ;

    return ;
}

#======================================================================================================================
# §function     runOnStore
# §state        private
#----------------------------------------------------------------------------------------------------------------------
# §syntax       pending
# §example      pending
#----------------------------------------------------------------------------------------------------------------------
# §description  pending
#----------------------------------------------------------------------------------------------------------------------
# §input        $Name | Description | type
#----------------------------------------------------------------------------------------------------------------------
# §return       $Name | Description | type
#======================================================================================================================
sub runOnStore {
    my $self = shift ;

    my $Console = $self->getConsole() ;
    my $Store = $self->{'ePages'}->getStore() ;
    $Console->info( "  Connecting to store '$Store' ...\n" ) ;
    eval {
        my $Arguments = $self->getArguments() ;
        my $ObjectPath = $Arguments->{'@'}->[1] // '/' ; 
        RunOnStore(
          'Store' => $Store,
          'Sub'   => sub {
            $self->{'ePages'}->connect() ;
            $self->executeCommand( 'cd', $ObjectPath ) ;
            $self->SUPER::_run() ;
          }
        ) ;
    };
    if ( $@ ) {
        $self->error( $@ ) ;
        $Console->output( "\n" ) ;
        $self->{'ePages'}->reset() ;
        $self->SUPER::_run() ;
    }

    return;
}

#======================================================================================================================
# §function     error
# §state        public
#----------------------------------------------------------------------------------------------------------------------
# §syntax       $Shell->error( $@ )
#----------------------------------------------------------------------------------------------------------------------
# §description  TODO
#----------------------------------------------------------------------------------------------------------------------
# §input        $ShowByeText | TODO | boolean
#======================================================================================================================
sub error {
    my $self = shift;

    my $Error = @_ ;
    
    if ( ExistsError() ) {
        $Error = GetError() ;
        $Error = sprintf( "[%s] '%s'", $Error->{'Code'}, $Error->{'Message'} ) ;
    }

    $self->getConsole()->error( $Error ) ;

    return ;
}

1;
